localstack/services/firehose/provider.py

```diff
@@ -0,0 +1,610 @@
+                    record_to_send["records"].append({"data": to_str(data)})
+                headers = {
+                    "Content-Type": "application/json",
+                }
+                try:
+                    requests.post(url, json=record_to_send, headers=headers)
+                except Exception as e:
+                    LOG.exception(f"Unable to put Firehose records to HTTP endpoint {url}.")
+                    raise e
+        return [
+            PutRecordBatchResponseEntry(RecordId=str(uuid.uuid4())) for _ in unprocessed_records
+        ]
+
+    def _add_missing_record_attributes(self, records: List[Dict]) -> None:
+        def _get_entry(obj, key):
+            return obj.get(key) or obj.get(first_char_to_lower(key))
+
+        for record in records:
+            if not _get_entry(record, "ApproximateArrivalTimestamp"):
+                record["ApproximateArrivalTimestamp"] = int(now_utc(millis=True))
+            if not _get_entry(record, "KinesisRecordMetadata"):
+                record["kinesisRecordMetadata"] = {
+                    "shardId": "shardId-000000000000",
+                    # not really documented what AWS is using internally - simply using a random UUID here
+                    "partitionKey": str(uuid.uuid4()),
+                    "approximateArrivalTimestamp": timestamp(
+                        float(_get_entry(record, "ApproximateArrivalTimestamp")) / 1000,
+                        format=TIMESTAMP_FORMAT_MICROS,
+                    ),
+                    "sequenceNumber": next_sequence_number(),
+                    "subsequenceNumber": "",
+                }
+
+    def _preprocess_records(self, processor: Dict, records: List[Record]) -> List[Dict]:
+        """Preprocess the list of records by calling the given processor (e.g., Lamnda function)."""
+        proc_type = processor.get("Type")
+        parameters = processor.get("Parameters", [])
+        parameters = {p["ParameterName"]: p["ParameterValue"] for p in parameters}
+        if proc_type == "Lambda":
+            lambda_arn = parameters.get("LambdaArn")
+            # TODO: add support for other parameters, e.g., NumberOfRetries, BufferSizeInMBs, BufferIntervalInSeconds, ...
+            client = aws_stack.connect_to_service("lambda")
+            records = keys_to_lower(records)
+            # Convert the record data to string (for json serialization)
+            for record in records:
+                if "data" in record:
+                    record["data"] = to_str(record["data"])
+                if "Data" in record:
+                    record["Data"] = to_str(record["Data"])
+            event = {"records": records}
+            event = to_bytes(json.dumps(event))
+            response = client.invoke(FunctionName=lambda_arn, Payload=event)
+            result = response.get("Payload").read()
+            result = json.loads(to_str(result))
+            records = result.get("records", []) if result else []
```

Comment:
Does this mean that if the Lambda function errors out or returns some invalid response that records we attempted to process are silently dropped?

Reply:
Yes, that's correct. However, I am not sure how the AWS service handles failures of processing lambdas (I just migrated the behavior as it was before).

Reply:
I'm not sure either actually, I've never had to use the Lambda integration 🤔 I'm guessing the records just get put into S3 like in other failure cases with Firehose. Might want to add a warning here.